clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -686,12 +686,31 @@ static Value *EmitSignBit(CodeGenFunction &CGF, Value *V) {
   return CGF.Builder.CreateICmpSLT(V, Zero);
 }
 
+/// Checks no arguments or results are passed indirectly in the ABI (i.e. via a
+/// hidden pointer). This is used to check annotating FP libcalls (that could
+/// set `errno`) with "int" TBAA metadata is safe. If any floating-point
+/// arguments are passed indirectly, setup for the call could be incorrectly
+/// optimized out.
+static bool HasNoIndirectArgumentsOrResults(CGFunctionInfo const &FnInfo) {
+  auto IsIndirect = [&](ABIArgInfo const &info) {
```

Comment:
May I ask why don't you determine whether the type of return and arguments is a pointer?  What's the difference?

Reply:
If you mean on the clang function declaration, they're not a pointer there (it's just how the ABI says to pass them).

If you mean why not check the LLVM function declaration, one reason is the pointers are opaque/untyped there, which may prevent setting the metadata on functions like `float frexpf(float, int*)` (which is a TODO in one of the tests).